docs/vendor-options.md
# Vendor Specific Options

DHCP from its inception was meant to be extensible with little effort. One of
the largest parts are vendor specific options provided to clients. With PG DHCP,
you can define custom options to be encapsulated within the vendor specific
information option (option 43).

## Declaring Option Types

Every option used in an option 43 segment must be define. The definition simply
needs the option code and data type.

```
decloption test-opt
    code 18
    type string
end
```

The above example declares a new vendor option named `test-opt`. All vendor names
must be unique within the configuration but they can be any string value. The
names are only used internally and not sent to the client.

The option code is 18 with a type string. All vendor options are type checked
just like normal options. Using the option is the same as using any other option.

```
network network1
    subnet 10.0.2.0/24
        range 10.0.2.10 10.0.2.200
        option router 10.0.2.1
        option test-opt "test"
        option vendor-options true
    end
end
```

The custom option is defined in the subnet with the value "test".

**NOTE:** The option `vendor-options` must be set to true in a subnet where you
want the vendor information to be sent. If `vendor-options` is not set or is
false, vendor options will NOT be sent. Also, make sure all vendor options are
defined BEFORE declaring `vendor-options` to true.

## Vendor Options Codes

Vendor codes can be any number between 1 and 254 inclusive. You will need to
consult the vendor documentation for which code any type to use.

## Vendor Option Types

The available types for vendor options are:

- `bool` - Boolean, true/false
- `address` - Single IPv4 address
- `address-list` - List of IPv4 addresses
- `string` - String
- `int8` - 8 bit number (0 - 255)
- `int16` - 16 bit number (0 - 65535)
- `int32` - 32 bit number (0 - 2147483647)


docs/appconf-settings.md
# Application Configuration

The application uses TOML for its configuration file.

## Available Settings

```toml
[logging]
Disabled = true         # Disable logging to a file
Level    = "debug"      # Logging level: debug, info, notice, warning, error, critical, alert, emergency, fatal
Path     = "dhcp.log"   # Path of log file

[Database]
Type     = "boltdb"      # Storage type "boltdb", "memory", "mysql", "pg"
Path     = "database.db" # Path to database for "boltdb"
Username = "root"        # Username for "mysql" and "pg"
Password = "password"    # Password for "mysql" and "pg"
Protocol = "tcp"         # Protocol for "mysql" and "pg"
Address  = "localhost"   # Address for "mysql" and "pg"
Port     = 3306          # Port for "mysql" and "pg"
Name     = "pg"          # Database name for "mysql" and "pg"

# Unless you have a VERY good reason, don't change these.
LeaseTable     = "lease"        # Lease table for "pg"
DeviceTable    = "device"       # Device table for "pg"
BlacklistTable = "blacklist"    # Blacklist table for "pg"

[leases]
DeleteAfter = "96h"     # Duration after which old leases are deleted, Go's time.Duration syntax

[server]
BlockBlacklisted = false            # Completely block blacklisted devices
NetworksFile     = "networks.conf"  # Path to network definition file
Workers          = 4                # Number of request workers

[management]
Address    = "0.0.0.0"      # IP address to expose management API
Port       = 8677           # Port to expose management API
AllowedIPs = ["10.2.3.5"]   # List of IP addresses that can access the management API
```

## Storage Options

### BoltDB

BoltDB is a file-based, blob storage that's typically much more performant than SQLite or other
file-based databases for blob storage. It's a pure Go implementation which means it's very efficient
and can easily be used cross-platform. This is a good choice for small installations or when running
as a stand-alone server.

### MySQL / MariaDB

Using a typical SQL database can make manual management much easier. BoltDB doesn't currently have a
canonical tool to manage Bolt databases. Also, only one process can use a Bolt database at a time.
Keep in mind that the server uses heavy internal caching especially for leases. Any manual changes to
leases while the server is running will be overwritten or at best completely ignored until the server
is restarted.

**Note**: The MySQL server must run in ANSI mode. This can achieved by running mysql with the `--ansi`
flag to editing the configuration file and adding `sql-mode = "ANSI"` to the `[mysqld]` section.

### PG (Packet Guardian)

This storage type is a modified version of the MySQL storage which allows using an existing Packet Guardian
database instead of having to migrate to a separate database on upgrade.

When using this storage, the management API is downgraded to limited, read-only functionality. Any calls to
alter a Device object will succeed but not do anything. This is because Devices are managed by the Packet
Guardian registration system and not the DHCP server.


internal/config/config.go
package config

import (
	"errors"
	"io/ioutil"
	"os"
	"runtime"
	"slices"
	"time"

	"github.com/naoina/toml"
	"github.com/packet-guardian/pg-dhcp/internal/utils"
)

type Config struct {
	sourceFile string
	Logging    *LoggingConfig
	Database   *DatabaseConfig
	Leases     *LeasesConfig
	Server     *ServerConfig
	Management *ManagementConfig
}

type LoggingConfig struct {
	Disabled bool
	Level    string
	Path     string
}

type DatabaseConfig struct {
	Type     string
	Path     string
	Username string
	Password string
	Protocol string
	Address  string
	Port     int
	Name     string

	LeaseTable     string
	DeviceTable    string
	BlacklistTable string
}

type LeasesConfig struct {
	DeleteAfter string // TODO: Run a job to clean up old leases
}

type ServerConfig struct {
	BlockBlacklisted bool
	NetworksFile     string
	Workers          int
}

type ManagementConfig struct {
	Address    string
	Port       int
	AllowedIPs []string
}

func FindConfigFile() string {
	if os.Getenv("PG_DHCP_CONFIG") != "" && utils.FileExists(os.Getenv("PG_DHCP_CONFIG")) {
		return os.Getenv("PG_DHCP_CONFIG")
	} else if utils.FileExists("./config.toml") {
		return "./config.toml"
	} else if utils.FileExists("/etc/pg-dhcp/config.toml") {
		return "/etc/pg-dhcp/config.toml"
	}
	return ""
}

func NewEmptyConfig() *Config {
	return &Config{}
}

func NewConfig(configFile string) (conf *Config, err error) {
	defer func() {
		if r := recover(); r != nil {
			switch x := r.(type) {
			case string:
				err = errors.New(x)
			case error:
				err = x
			default:
				err = errors.New("Unknown panic")
			}
		}
	}()

	if configFile == "" {
		configFile = "config.toml"
	}

	f, err := os.Open(configFile)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	buf, err := ioutil.ReadAll(f)
	if err != nil {
		return nil, err
	}
	var con Config
	if err := toml.Unmarshal(buf, &con); err != nil {
		return nil, err
	}
	con.sourceFile = configFile
	return setSensibleDefaults(&con)
}

func setSensibleDefaults(c *Config) (*Config, error) {
	// Anything not set here implies its zero value is the default

	// Ensure all substructs exists
	if c.Logging == nil {
		c.Logging = &LoggingConfig{}
	}
	if c.Logging == nil {
		c.Logging = &LoggingConfig{}
	}
	if c.Database == nil {
		c.Database = &DatabaseConfig{}
	}
	if c.Leases == nil {
		c.Leases = &LeasesConfig{}
	}
	if c.Server == nil {
		c.Server = &ServerConfig{}
	}
	if c.Management == nil {
		c.Management = &ManagementConfig{}
	}

	// Logging
	c.Logging.Level = setStringOrDefault(c.Logging.Level, "notice")
	c.Logging.Path = setStringOrDefault(c.Logging.Path, "logs/pg.log")

	// Database
	c.Database.Type = setStringOrDefault(c.Database.Type, "boltdb")
	c.Database.Path = setStringOrDefault(c.Database.Path, "database.db")
	c.Database.Username = setStringOrDefault(c.Database.Username, "root")
	c.Database.Password = setStringOrDefault(c.Database.Password, "password")
	c.Database.Protocol = setStringOrDefault(c.Database.Protocol, "tcp")
	c.Database.Address = setStringOrDefault(c.Database.Address, "localhost")
	c.Database.Port = setIntOrDefault(c.Database.Port, 3306)
	c.Database.Name = setStringOrDefault(c.Database.Name, "pg")

	c.Database.LeaseTable = setStringOrDefault(c.Database.LeaseTable, "lease")
	c.Database.DeviceTable = setStringOrDefault(c.Database.DeviceTable, "device")
	c.Database.BlacklistTable = setStringOrDefault(c.Database.BlacklistTable, "blacklist")

	// Leases
	c.Leases.DeleteAfter = setStringOrDefault(c.Leases.DeleteAfter, "96h")
	if _, err := time.ParseDuration(c.Leases.DeleteAfter); err != nil {
		c.Leases.DeleteAfter = "96h"
	}

	// DHCP
	c.Server.NetworksFile = setStringOrDefault(c.Server.NetworksFile, "/etc/pg-dhcp/dhcp.conf")
	c.Server.Workers = setIntOrDefault(c.Server.Workers, runtime.GOMAXPROCS(0))

	// Management
	c.Management.Address = setStringOrDefault(c.Management.Address, "0.0.0.0")
	c.Management.Port = setIntOrDefault(c.Management.Port, 8677)
	if c.Management.AllowedIPs != nil {
		if slices.Contains(c.Management.AllowedIPs, "0.0.0.0") {
			// 0.0.0.0 matches every address, setting this to nil is as if it was never set.
			c.Management.AllowedIPs = nil
		}
	}

	return c, nil
}

// Given string s, if it is empty, return v else return s.
func setStringOrDefault(s, v string) string {
	if s == "" {
		return v
	}
	return s
}

// Given int s, if it is zero, return v else return s.
func setIntOrDefault(s, v int) int {
	if s == 0 {
		return v
	}
	return s
}
